fs/operations/resume.go

```diff
@@ -0,0 +1,178 @@
+		fingerprint := fs.Fingerprint(ctx, src, true)
+		data, err := marshalResumeJSON(ctx, fingerprint, ID, hashName, hashState)
+		if err != nil {
+			return errors.Wrapf(err, "failed to marshal data JSON")
+		}
+		if len(data) < int(ci.MaxResumeCacheSize) {
+			// Each remote will have its own dir for cached resume files
+			dirPath := filepath.Join(rootCacheDir, f.Name(), f.Root())
+			err = os.MkdirAll(dirPath, os.ModePerm)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create cache directory %v", dirPath)
+			}
+			// Write resume data to disk
+			cachePath := filepath.Join(dirPath, remote)
+			cacheFile, err := os.Create(cachePath)
+			if err != nil {
+				return errors.Wrapf(err, "failed to create cache file %v", cachePath)
+			}
+			defer func() {
+				_ = cacheFile.Close()
+			}()
+			_, errWrite := cacheFile.Write(data)
+			if err != nil {
+				return errors.Wrapf(errWrite, "failed to write JSON to file")
+			}
+		}
+		if !cacheCleaned {
+			if err := cleanCache(ctx, rootCacheDir); err != nil {
+				return errors.Wrapf(err, "failed to clean resume cache")
+			}
+		}
+		cacheCleaned = true
+		return nil
+	}
+}
+
+// cleanCache checks the size of the resume cache and removes the oldest resume files if more than limit
+func cleanCache(ctx context.Context, rootCacheDir string) error {
+	ci := fs.GetConfig(ctx)
+	var paths []string
+	pathsWithInfo := make(map[string]os.FileInfo)
+	totalCacheSize := int64(0)
+	walkErr := filepath.Walk(rootCacheDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				// Empty subdirectories in the resume cache dir can be removed
+				removeErr := os.Remove(path)
+				if err != nil && !os.IsNotExist(removeErr) {
+					return errors.Wrapf(err, "failed to remove empty subdirectory: %s", path)
+				}
+				return nil
+			}
+			paths = append(paths, path)
+			pathsWithInfo[path] = info
+			totalCacheSize += info.Size()
+			return nil
+		})
+	if walkErr != nil {
+		return errors.Wrapf(walkErr, "error walking through cache when cleaning cache dir")
+	}
+	if totalCacheSize > int64(ci.MaxResumeCacheSize) {
+		sort.Slice(paths, func(i, j int) bool {
+			return pathsWithInfo[paths[i]].ModTime().Before(pathsWithInfo[paths[j]].ModTime())
+		})
+		for _, p := range paths {
+			if totalCacheSize < int64(ci.MaxResumeCacheSize) {
+				break
+			}
+			if err := os.Remove(p); err != nil {
+				return errors.Wrapf(err, "error removing oldest cache file: %s", p)
+			}
+			totalCacheSize -= pathsWithInfo[p].Size()
+			fs.Debugf(p, "Successfully removed oldest cache file")
+		}
+	}
+	return nil
+}
+
+// readResumeCache checks to see if a resume ID has been cached for the source object.
+// If it finds one it returns it along with true to signal a resume can be attempted
+func readResumeCache(ctx context.Context, f fs.Fs, src fs.Object, cacheName string) (resumeID, hashName, hashState string, attemptResume bool) {
```

Comment:
This implementation was written with the chunker in mind, but not all backends need to manage uploads as a set of files.

Assuming that there is a cache folder available forces backend resumable uploads into a specific method of managing the uploads.

For example, if I wanted to implement resumable uploads using the multipart upload functionality of the AWS S3 API, then there are no objects created in the backend until the upload is complete.

Reply:
See above

Reply:
The implementation is generic enough to handle every suppirted backend.

Reply:
The point isn't that some backends are not capable of managing a cache, it is that this forces backends to implement a cache when they might already have better support for resumable uploads via a different mechanism.

Reply:
> this forces backends to implement a cache

No. They will just return empty hash.

Reply:
> resumable uploads via a different mechanism

If eg. S3 needs other data to resume an upload, that other data will be added to the signature. If the signature grows above 3-4 parameters, we can factor it and pass a structure.

Reply:
> the implementation should just be composable rather than ignorable parameters.

I'd prefer the latter. This is discussible.